tests/unit/lms/validation/blackboard_test.py

```diff
@@ -0,0 +1,122 @@
+                            "unknown_field": "this_should_be_included",
+                        },
+                        {
+                            "id": "_7851_2",
+                            "name": "File_2.pdf",
+                            "modified": "2021-05-02T15:39:51.000Z",
+                        },
+                        {
+                            "id": "_7851_3",
+                            "name": "File_3.pdf",
+                            "modified": "2021-04-02T15:39:51.000Z",
+                        },
+                    ],
+                },
+                # The output is just the list of files, directly.
+                [
+                    {
+                        "id": "_7851_1",
+                        "name": "File_1.pdf",
+                        "modified": "2021-06-02T15:39:51.000Z",
+                        "unknown_field": "this_should_be_included",
+                    },
+                    {
+                        "id": "_7851_2",
+                        "name": "File_2.pdf",
+                        "modified": "2021-05-02T15:39:51.000Z",
+                    },
+                    {
+                        "id": "_7851_3",
+                        "name": "File_3.pdf",
+                        "modified": "2021-04-02T15:39:51.000Z",
+                    },
+                ],
+            )
+        ],
+    )
+    def test_valid(self, input_, output):
+        schema = BlackboardListFilesSchema(
+            factories.requests.Response(json_data=input_)
+        )
+
+        result = schema.parse()
+
+        assert result == output
+
+    @pytest.mark.parametrize(
+        "json_data,messages",
+        [
+            # No top-level "results" list.
+            ({}, {"results": ["Missing data for required field."]}),
+            # The top-level "results" list is not a list.
+            ({"results": None}, {"results": ["Field may not be null."]}),
+            # A file in the "results" list is missing required fields.
+            (
+                {
+                    "results": [
+                        {
+                            "foo": "bar",
+                        }
+                    ]
+                },
+                {
+                    "results": {
+                        0: {
+                            "id": ["Missing data for required field."],
+                            "modified": ["Missing data for required field."],
+                            "name": ["Missing data for required field."],
+                        }
+                    }
+                },
+            ),
+            # A file in the "results" list has invalid values.
+            (
+                {
+                    "results": [
+                        {
+                            "id": 23,
+                            "name": None,
+                            "modified": False,
+                        },
+                    ]
+                },
```

Comment:
It might make a more unit-y test to separate the tests into:

 * Those where the parent structure is wrong
 * Those where elements of a specific file are wrong
 
That would allow you to have a "good" file and then modify each field individually with something like this:

```python
@pytest.mark.parametrize("field", ("id", "modified", "name"))
@pytest.mark.parametrize("bad_value", (MISSING, None, 1234))
def test_invalid_field(self, json_data, messages):        
    bad_file = {"id": "string",  "modified": "string",  "name": "string"}
    if value is MISSING:
        bad_file.pop(field)
    else:
        bad_file[field] = value

    schema = BlackboardListFilesSchema(
        factories.requests.Response(json_data={"results": [bad_file]})
    )        
    with pytest.raises(ValidationError) as exc_info: 
        schema.parse()
```

Reply:
Done.

I didn't end up doing _exactly_ what's suggested here but something very similar. I split up the single `test_invalid()` test (which the big `parametrize()`) into separate tests and I implemented the thing where the tests contain just a single valid API response dict and mutate that dict to make it invalid in different ways.

Here's an example of what one of the tests looks like now:

```python
@pytest.mark.parametrize("missing_field", ["id", "modified", "name"])
def test_it_raises_if_a_required_field_is_missing(self, json_data, assert_raises, missing_field):
    del json_data["results"][0][missing_field]

    assert_raises()
```

I think the change to using a valid `json_data` fixture and mutating it also makes it unnecessary to assert about the actual error messages (see other comment above about that)